// countup.js-angular2/esm5/lib/count-up.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, Output, HostListener, EventEmitter } from '@angular/core';
import * as CountUp from 'countup.js';
var CountUpDirective = /** @class */ (function () {
    function CountUpDirective(el) {
        this.el = el;
        // on-complete event emitter
        this.complete = new EventEmitter();
    }
    // Re-animate if preference is set.
    // Re-animate if preference is set.
    /**
     * @return {?}
     */
    CountUpDirective.prototype.onClick = 
    // Re-animate if preference is set.
    /**
     * @return {?}
     */
    function () {
        if (this.reanimateOnClick) {
            this.animate();
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    CountUpDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.endVal && changes.endVal.currentValue) {
            this.countUp = this.createCountUp();
            this.animate();
        }
    };
    /**
     * @return {?}
     */
    CountUpDirective.prototype.createCountUp = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var start = this.startVal || 0;
        /** @type {?} */
        var duration = this.duration || 2;
        /** @type {?} */
        var decimals = this.decimals || 0;
        if (!this.duration) {
            this.duration = duration;
        }
        // construct countUp
        /** @type {?} */
        var countUp = new CountUp(this.el.nativeElement, start, this.endVal, decimals, duration, this.options);
        /** @type {?} */
        var diff = Math.abs(this.endVal - start);
        // make easing smoother for large numbers
        if (diff > 999) {
            /** @type {?} */
            var up = (this.endVal > start) ? -1 : 1;
            countUp = new CountUp(this.el.nativeElement, start, this.endVal + (up * 100), decimals, duration / 2, this.options);
        }
        return countUp;
    };
    /**
     * @return {?}
     */
    CountUpDirective.prototype.animate = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.countUp.reset();
        if (this.endVal > 999) {
            this.countUp.start(function () { return _this.countUp.update(_this.endVal); });
            setTimeout(function () {
                _this.complete.emit();
            }, this.duration * 1000);
        }
        else {
            this.countUp.start(function () {
                _this.complete.emit();
            });
        }
    };
    CountUpDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[countUp]'
                },] }
    ];
    CountUpDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    CountUpDirective.propDecorators = {
        options: [{ type: Input, args: ['countUp',] }],
        startVal: [{ type: Input }],
        endVal: [{ type: Input }],
        duration: [{ type: Input }],
        decimals: [{ type: Input }],
        reanimateOnClick: [{ type: Input }],
        onClick: [{ type: HostListener, args: ['click',] }],
        complete: [{ type: Output }]
    };
    return CountUpDirective;
}());
export { CountUpDirective };
if (false) {
    /** @type {?} */
    CountUpDirective.prototype.countUp;
    /** @type {?} */
    CountUpDirective.prototype.options;
    /** @type {?} */
    CountUpDirective.prototype.startVal;
    /** @type {?} */
    CountUpDirective.prototype.endVal;
    /** @type {?} */
    CountUpDirective.prototype.duration;
    /** @type {?} */
    CountUpDirective.prototype.decimals;
    /** @type {?} */
    CountUpDirective.prototype.reanimateOnClick;
    /** @type {?} */
    CountUpDirective.prototype.complete;
    /** @type {?} */
    CountUpDirective.prototype.el;
}
