// countup.js-angular2/bundles/countup.js-angular2.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('countup.js')) :
    typeof define === 'function' && define.amd ? define('countup.js-angular2', ['exports', '@angular/core', 'countup.js'], factory) :
    (factory((global.countup = global.countup || {}, global.countup['js-angular2'] = {}),global.ng.core,global.CountUp));
}(this, (function (exports,core,CountUp) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    var CountUpDirective = /** @class */ (function () {
        function CountUpDirective(el) {
            this.el = el;
            // on-complete event emitter
            this.complete = new core.EventEmitter();
        }
        // Re-animate if preference is set.
        // Re-animate if preference is set.
        /**
         * @return {?}
         */
        CountUpDirective.prototype.onClick =
            // Re-animate if preference is set.
            /**
             * @return {?}
             */
            function () {
                if (this.reanimateOnClick) {
                    this.animate();
                }
            };
        /**
         * @param {?} changes
         * @return {?}
         */
        CountUpDirective.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                if (changes.endVal && changes.endVal.currentValue) {
                    this.countUp = this.createCountUp();
                    this.animate();
                }
            };
        /**
         * @return {?}
         */
        CountUpDirective.prototype.createCountUp = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var start = this.startVal || 0;
                /** @type {?} */
                var duration = this.duration || 2;
                /** @type {?} */
                var decimals = this.decimals || 0;
                if (!this.duration) {
                    this.duration = duration;
                }
                // construct countUp
                /** @type {?} */
                var countUp = new CountUp(this.el.nativeElement, start, this.endVal, decimals, duration, this.options);
                /** @type {?} */
                var diff = Math.abs(this.endVal - start);
                // make easing smoother for large numbers
                if (diff > 999) {
                    /** @type {?} */
                    var up = (this.endVal > start) ? -1 : 1;
                    countUp = new CountUp(this.el.nativeElement, start, this.endVal + (up * 100), decimals, duration / 2, this.options);
                }
                return countUp;
            };
        /**
         * @return {?}
         */
        CountUpDirective.prototype.animate = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.countUp.reset();
                if (this.endVal > 999) {
                    this.countUp.start(function () { return _this.countUp.update(_this.endVal); });
                    setTimeout(function () {
                        _this.complete.emit();
                    }, this.duration * 1000);
                }
                else {
                    this.countUp.start(function () {
                        _this.complete.emit();
                    });
                }
            };
        CountUpDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[countUp]'
                    },] }
        ];
        CountUpDirective.ctorParameters = function () {
            return [
                { type: core.ElementRef }
            ];
        };
        CountUpDirective.propDecorators = {
            options: [{ type: core.Input, args: ['countUp',] }],
            startVal: [{ type: core.Input }],
            endVal: [{ type: core.Input }],
            duration: [{ type: core.Input }],
            decimals: [{ type: core.Input }],
            reanimateOnClick: [{ type: core.Input }],
            onClick: [{ type: core.HostListener, args: ['click',] }],
            complete: [{ type: core.Output }]
        };
        return CountUpDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    var CountUpModule = /** @class */ (function () {
        function CountUpModule() {
        }
        CountUpModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [],
                        declarations: [CountUpDirective],
                        exports: [CountUpDirective]
                    },] }
        ];
        return CountUpModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */

    exports.CountUpModule = CountUpModule;
    exports.ɵa = CountUpDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
