// countup.js-angular2/esm2015/lib/count-up.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, Output, HostListener, EventEmitter } from '@angular/core';
import * as CountUp from 'countup.js';
export class CountUpDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
        // on-complete event emitter
        this.complete = new EventEmitter();
    }
    // Re-animate if preference is set.
    /**
     * @return {?}
     */
    onClick() {
        if (this.reanimateOnClick) {
            this.animate();
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.endVal && changes.endVal.currentValue) {
            this.countUp = this.createCountUp();
            this.animate();
        }
    }
    /**
     * @return {?}
     */
    createCountUp() {
        /** @type {?} */
        const start = this.startVal || 0;
        /** @type {?} */
        const duration = this.duration || 2;
        /** @type {?} */
        const decimals = this.decimals || 0;
        if (!this.duration) {
            this.duration = duration;
        }
        // construct countUp
        /** @type {?} */
        let countUp = new CountUp(this.el.nativeElement, start, this.endVal, decimals, duration, this.options);
        /** @type {?} */
        const diff = Math.abs(this.endVal - start);
        // make easing smoother for large numbers
        if (diff > 999) {
            /** @type {?} */
            const up = (this.endVal > start) ? -1 : 1;
            countUp = new CountUp(this.el.nativeElement, start, this.endVal + (up * 100), decimals, duration / 2, this.options);
        }
        return countUp;
    }
    /**
     * @return {?}
     */
    animate() {
        this.countUp.reset();
        if (this.endVal > 999) {
            this.countUp.start(() => this.countUp.update(this.endVal));
            setTimeout(() => {
                this.complete.emit();
            }, this.duration * 1000);
        }
        else {
            this.countUp.start(() => {
                this.complete.emit();
            });
        }
    }
}
CountUpDirective.decorators = [
    { type: Directive, args: [{
                selector: '[countUp]'
            },] }
];
CountUpDirective.ctorParameters = () => [
    { type: ElementRef }
];
CountUpDirective.propDecorators = {
    options: [{ type: Input, args: ['countUp',] }],
    startVal: [{ type: Input }],
    endVal: [{ type: Input }],
    duration: [{ type: Input }],
    decimals: [{ type: Input }],
    reanimateOnClick: [{ type: Input }],
    onClick: [{ type: HostListener, args: ['click',] }],
    complete: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    CountUpDirective.prototype.countUp;
    /** @type {?} */
    CountUpDirective.prototype.options;
    /** @type {?} */
    CountUpDirective.prototype.startVal;
    /** @type {?} */
    CountUpDirective.prototype.endVal;
    /** @type {?} */
    CountUpDirective.prototype.duration;
    /** @type {?} */
    CountUpDirective.prototype.decimals;
    /** @type {?} */
    CountUpDirective.prototype.reanimateOnClick;
    /** @type {?} */
    CountUpDirective.prototype.complete;
    /** @type {?} */
    CountUpDirective.prototype.el;
}
